File: code/python/function3.py
from itertools import groupby
from math import floor


def function3(l, n):
    assert len(l) > 0 and n > 0
    minimum = min(l)
    maximum = max(l)
    d = (maximum - minimum) / n
    values = [min(n - 1, (floor((l[i] - minimum) / d))) for i in range(len(l))]
    dictionary = dict(
        [(key, len(list(group))) for key, group in groupby(sorted(values))]
    )
    return [(minimum + (i + 0.5) * d, dictionary.get(i, 0)) for i in range(0, n, 1)]

File: code/scala/function3.scala
def function3(seq: Seq[Double], n: Int): Seq[(Double, Int)] = {
    assert(seq.length > 0 && n > 0)

    val min: Double = seq.min
    val max: Double = seq.max

    val d: Double = (max - min) / n
    
    val base: Seq[(Int, Int)] = (0 until n).map(e => (e -> 0))
    val adder: Seq[(Int, Int)] = seq.map(e => math.min(n - 1, math.floor((e - min)/d).toInt)).groupBy(identity).toSeq.map{
    case (key, list) => key -> list.size
    }

    (base ++ adder).groupBy(_._1).toSeq.map(e => (min + ((2*e._1+1)/2.0)* d) -> e._2.map(_._2).sum).sortBy(_._1)
    
}